### DISCUSSION TITLE
Share encrypted session across multiple apps

### DISCUSSION THREAD

**Discussion by muellerj** [jeremyevans/roda#226] [General]
Hi there,

I would like to share an (encrypted) session across several app living on separate subdomains of the same domain. E.g. `foo.example.com` and `bar.example.com` both have the following setup:

```
  use RodaSessionMiddleware,
    key: "foo.session",
    secret: MY_SHARED_SECRET_FOR_BOTH_ENDPOINTS,
    cookie_options: { domain: "example.com" }
```

`foo.example.com` sets `session["value"] = 42` and then `bar.example.com` (having been set up with the same session secret and the same key) has access to `session["value"]`.

I have tried to pass `cookie_options: { domain: "example.com" }` or `cookie_options: { domain: ".example.com" }` to `RodaSessionMiddleware`, but with no success so far. 

Apparently, something similar can be done with `Rack::Session::Cookie`: https://github.com/redis-store/redis-store/issues/108

Can anyone give me a pointer?

**Comment by muellerj** [comment]
Nevermind, I think I managed to get it working. There is no need for the middleware, a configuration like so allows both subdomains to access the session:

```
  plugin :sessions,
    key: "foo.session",
    secret: MY_SHARED_SECRET_FOR_BOTH_ENDPOINTS,
    cookie_options: { domain: "example.com" }
```

**Comment by muellerj** [comment]
One more cotcha in case someone reads this: Clearing the session across multiple subdomains did not work as expected (the session persists, even if one of the endpoints explicitly clears it). I had to do something like this to manually delete the session cookie:

```
plugin :cookies, domain: "example.com"
after_logout do
  response.delete_cookie "foo.session"
end
```

**Comment by jeremyevans** [comment]
Clearing the session (`session.clear`) should delete the session cookie (Roda calls `Rack::Utils.delete_cookie_header!` when persisting an empty session, if a session cookie was provided with the request).  Can you put together a simple example showing the problem?

**Comment by muellerj** [comment]
Sure. I’ve so far tested this only on the live domain but will try to put something together tonight.

**Comment by muellerj** [comment]
I'm having trouble reproducing this without a reverse-proxy setup with subdomains. The following minimal example works as expected (setting/resetting in one app is reflected correctly in the other):

``` ruby
# app1.rb

require "roda"

SHARED_SECRET = ENV.fetch("SHARED_SECRET", "x" * 64)

class TestApp1 < Roda

  plugin :sessions,
    secret: SHARED_SECRET,
    key: "roda-test.session",
    cookie_options: { domain: "localhost" }

  route do |r|
    r.root do
      <<~EOS
      #{session["value"].inspect}
      <a href="/set">set</a> |
      <a href="/reset">reset</a>
      EOS
    end
    r.get "set" do
      session["value"] = 1
      r.redirect "/"
    end
    r.get "reset" do
      session.clear
      r.redirect "/"
    end
  end
  freeze
end

if $PROGRAM_NAME == __FILE__
  require "rack"
  Rack::Server.start(app: TestApp1, Port: 9292)
end
```

``` ruby
# app2.rb

require "roda"

SHARED_SECRET = ENV.fetch("SHARED_SECRET", "x" * 64)

class TestApp1 < Roda

  plugin :sessions,
    secret: SHARED_SECRET,
    key: "roda-test.session",
    cookie_options: { domain: "localhost" }

  route do |r|
    r.root do
      <<~EOS
      #{session["value"].inspect}
      <a href="/set">set</a> |
      <a href="/reset">reset</a>
      EOS
    end
    r.get "set" do
      session["value"] = 1
      r.redirect "/"
    end
    r.get "reset" do
      session.clear
      r.redirect "/"
    end
  end
  freeze
end

if $PROGRAM_NAME == __FILE__
  require "rack"
  Rack::Server.start(app: TestApp1, Port: 9293)
end
```

**Comment by jeremyevans** [comment]
Thank you for testing!  If you figure out what is causing the issue and it is due to a problem in Roda, please let me know so I can fix it.

**Comment by muellerj** [comment]
Just to follow up - I was not able to reproduce the error concerning session clearance mentioned in https://github.com/jeremyevans/roda/discussions/226#discussioncomment-1821440. Using a setup like

```
  plugin :sessions,
    key: "foo.session",
    secret: MY_SHARED_SECRET_FOR_BOTH_ENDPOINTS,
    cookie_options: { domain: "example.com" }
```

seems to work well and covers my initial requirement.
